tests/unit/s2n_ktls_io_test.c

```diff
@@ -286,5 +308,212 @@ int main(int argc, char **argv)
+            EXPECT_EQUAL(bytes_written, S2N_TEST_TO_SEND);
+
+            uint8_t recv_buf[S2N_TLS_MAXIMUM_FRAGMENT_LENGTH] = { 0 };
+            uint8_t recv_record_type = 0;
+            size_t bytes_read = 0;
+            EXPECT_OK(s2n_ktls_recvmsg(client, &recv_record_type, recv_buf, S2N_TEST_TO_SEND, &blocked, &bytes_read));
+            EXPECT_BYTEARRAY_EQUAL(test_data, recv_buf, bytes_read);
+            EXPECT_EQUAL(bytes_read, bytes_written);
+
+            EXPECT_EQUAL(io_pair.client_in.sendmsg_invoked_count, 1);
+            EXPECT_EQUAL(io_pair.client_in.recvmsg_invoked_count, 1);
+        };
+
+        /* Simulate blocked and handle a S2N_ERR_IO_BLOCKED error */
+        {
+            DEFER_CLEANUP(struct s2n_connection *server = s2n_connection_new(S2N_SERVER),
+                    s2n_connection_ptr_free);
+            DEFER_CLEANUP(struct s2n_connection *client = s2n_connection_new(S2N_CLIENT),
+                    s2n_connection_ptr_free);
+            DEFER_CLEANUP(struct s2n_test_ktls_io_stuffer_pair io_pair = { 0 },
+                    s2n_ktls_io_stuffer_pair_free);
+            EXPECT_OK(s2n_test_init_ktls_io_stuffer(server, client, &io_pair));
+
+            uint8_t recv_buf[S2N_TLS_MAXIMUM_FRAGMENT_LENGTH] = { 0 };
+            s2n_blocked_status blocked = S2N_NOT_BLOCKED;
+            size_t blocked_invoked_count = 5;
+            uint8_t recv_record_type = 0;
+            size_t bytes_read = 0;
+            /* recv should block since there is no data */
+            for (size_t i = 0; i < blocked_invoked_count; i++) {
+                EXPECT_ERROR_WITH_ERRNO(
+                        s2n_ktls_recvmsg(client, &recv_record_type, recv_buf, S2N_TEST_TO_SEND, &blocked, &bytes_read),
+                        S2N_ERR_IO_BLOCKED);
+                EXPECT_EQUAL(blocked, S2N_BLOCKED_ON_READ);
+            }
+
+            /* send data to unblock */
+            struct iovec msg_iov = { .iov_base = test_data, .iov_len = S2N_TEST_TO_SEND };
+            size_t bytes_written = 0;
+            EXPECT_OK(s2n_ktls_sendmsg(server, test_record_type, &msg_iov, 1, &blocked, &bytes_written));
+            EXPECT_EQUAL(bytes_written, S2N_TEST_TO_SEND);
+
+            EXPECT_OK(s2n_ktls_recvmsg(client, &recv_record_type, recv_buf, S2N_TEST_TO_SEND, &blocked, &bytes_read));
+            EXPECT_BYTEARRAY_EQUAL(test_data, recv_buf, bytes_read);
+            EXPECT_EQUAL(bytes_read, bytes_written);
+
+            /* recv should block again since we have read all the data */
+            for (size_t i = 0; i < blocked_invoked_count; i++) {
+                EXPECT_ERROR_WITH_ERRNO(
+                        s2n_ktls_recvmsg(client, &recv_record_type, recv_buf, S2N_TEST_TO_SEND, &blocked, &bytes_read),
+                        S2N_ERR_IO_BLOCKED);
+                EXPECT_EQUAL(blocked, S2N_BLOCKED_ON_READ);
+            }
+
+            EXPECT_EQUAL(io_pair.client_in.recvmsg_invoked_count, (blocked_invoked_count * 2) + 1);
```

Comment:
Why is this times 2?

Reply:
I call and test blocked in a loop twice:
- once before any data was sent
- once after we have read all the data available

Reply:
Probably worth a comment